### DISCUSSION TITLE
[V3] "Can't set model as property if it hasn't been persisted yet" when component deletes model

### DISCUSSION THREAD

**Discussion by riasvdv** [livewire/livewire#6120] [7. Bugs (archived)]
### Livewire version

v3.0.0-beta.5

### Which PHP version are you using?

PHP 8.2

### Repository URL

https://github.com/riasvdv/livewire-delete-self

### Steps To Reproduce

1. Seed the database
2. See 10 example comments
3. Press "Delete"

This behaviour works in v2 and it dispatches the `delete` event to the parent component, which re-renders and removes the row.

In V3 the following exception happens: https://flareapp.io/share/A7EoxRLm

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by romansh** [comment]
Have you tried registering the event listener in this way?

#[On('delete')]

**Reply by riasvdv** [reply]
Registering the event listener is unrelated to the exception, as the exception happens either way.

**Comment by joshhanley** [comment]
@riasvdv thanks for reporting and the example repo! I will pull it down and have a look when I get the chance. Assigning models to public properties has changed a bit since V2, so there will be some changes. I suspect the child is re-rendering when it gets deleted and can't find the model, hence the error when I would actually expect the child component to be removed. 

One thing, could you try this - can you try clearing the property on delete, after the `$this->comment->delete();` add `$this->comment = null;` and update the type on the property to accept `null` and let me know if that works?

Thanks!

**Reply by riasvdv** [reply]
Setting it to `null` still causes the component to render and causes an exception on the property (https://flareapp.io/share/17xDzoe7) , it seems like the parent component is not refreshing in time for it to be destroyed before it re-renders.

I can update the view to handle a `null` model and not render anything, which then refreshes correctly and removes the component, but that feels like a lot of extra work to make it work when it "just worked" in V2

**Reply by riasvdv** [reply]
I've created a PR with a failing test in the meantime https://github.com/livewire/livewire/pull/6272

**Reply by ithuis** [reply]
this works for me

```
$order = $this->order;
$this->order = null;
$order->delete();
```

**Reply by chuckrincon** [reply]
The @ithuis solution also worked for me.

**Comment by samloft** [comment]
Is this something that is considered a bug in V3 or expected behaviour? As in V2 I do things like the below **alot** in our applications.

![image](https://github.com/livewire/livewire/assets/19821958/69fb3b74-afd8-4b88-82c8-7604548a139c)

And now when editCompany is called I get the same **Can't set model as property if it hasn't been persisted yet** error.

**Reply by samloft** [reply]
Sorry, should have probably also mentioned.

This means when calling store methods I can just do

`$this->edit->save()`

**Comment by hayespdx** [comment]
I get the same error after upgrading, but not after deleting models. My Livewire components access arrays of data from my API, not Laravel models directly. The components then create differently-named models with the API data, so I can take advantage of the features of Laravel's Model class, e.g. accessors and mutators. I can then do something like {{ $image->url }} in my view components. So the models Livewire is working with are never persisted; rather, I send an API request to create and update the actual models. Can I no longer do this? :(

**Reply by samloft** [reply]
In your livewire.php config have you tried setting

legacy_model_binding to true?

This got me up and running again.

**Reply by hayespdx** [reply]
That's a great idea - I'm so glad it worked for you! Thanks for letting me know about it. Before I saw your reply I discovered computed properties, which is mostly working... but I'll keep your solution in mind as a fallback if I can't get it all the way there. Changing that setting now causes an error about MorphMaps, for whatever reason, so I'll continue in my direction for now.

**Comment by wizzymore** [comment]
I don't know if this is a problem with Livewire, I think it's more of how you approach components and their state.  
If you want to render a comment in it's own component i don't think the best way to delete it is in the component itself as it will still want to re-render after it is deleted and this is not practical.

The be approach would be to have the parent component `Comments` control the deletion of the specific comment through an event.  
You might have a more advanced use case than what you are showing us, still, not everything needs to be a livewire component, and for the example you provided it only adds complexity not simplicity.

This code works:

```php
    // Comment.php
    
    public function delete()
    {
        $this->dispatch('deleteComment', $this->comment->id);
    }
```

```php
    // Comments.php
    protected $listeners = [
        'deleteComment' => 'deleteComment',
    ];

    public function deleteComment(Comment $comment)
    {
        $comment->delete();
    }
```

**Reply by flatcapco** [reply]
This gets a bit messy when your user flow is parent component (table of models) then the remove button is a nested component to handle a confirmation popup with some other logic and then you have to pass it back to the parent component.

Its a lot nicer handling it within the popup modal or you're passing it around twice - are there any more thoughts on this?

**Reply by wizzymore** [reply]
Can you explain further or provide an example?

**Reply by flatcapco** [reply]
No problem.

I have a booking summary table where guests can remove any item from the table they'd like. I have a column by each item with a remove button that dispatches the following:

`wire:click="$dispatch('openModal', { component: 'remove-booking-item', arguments: { booking_id: {{ $booking->id }}, item_id: {{ $item->id }}  } })"`

I have a modal component RemoveBookingItem that extends a global ModalComponent
in here I handle my specific business logic before allowing the delete... if that is all good then the guest gets the delete button and on click the modal closes.

`    
public function remove(): void
    {
        $this->dispatch('closeModal');
        $this->item->delete();
    }`

With the legacy mode on it works - the item is removed from the booking summary table in the parent component.. however I'd like to keep this disabled and go forwards with l3 logic.

If I understood correctly your suggestion would mean that I'd have to dispatch another command and then listen for that on the parent component and then have my deletion logic on there?

Its not really convienient for me because there are many places that handle these items in different ways and it'd mean a lot of passing about

**Reply by wizzymore** [reply]
The problem is that the Component that holds the data on which you iterate to display the individual components needs to be aware of the fact that one of those individual components should be removed from the dom on the next render so it doesn't hydrate it again, because Livewire will not be able to hydrate the model as it is deleted and the above error will be thrown.

One quick thing I have in mind is having the lone component that handles the "row" to be able to accept a null value in place of the model, so that you can call the `delete()` method on the model and then set it's value to null and render an empty component when the model value is null. You still need to notify the parent about the change so it can re-render the list, but, in this way Livewire can successfully hydrate.
This is more of a trick than a solution.

I find it really painful to be honest to work with models and the hydration that livewire uses. Most often it does things that I do not want it to do, or does more queries then it needs to do and I want granular control over this. For me, at least, it's easier to work with plain arrays of data instead of models and when I want to control the DB data itself I do the queries as needed.

Keep in mind I am very biased and I don't like working with ORMs so it might not be the best solution for you.

**Reply by flatcapco** [reply]
Thanks for your reply - i think your thoughts on how to handle the row is a good enough solution for now until I've had time to sit down and handle it better. I appreciate it

**Comment by ccssmania** [comment]
I have a similar issue, but not when deleting the model from the component but when deleting from outside instead. I'm using a component that shows the status of a document that is being scanned. If the scan doesn't find any threat, the document is updated OK, and the component shows the status as scanned, but when a threat is found, the document gets deleted.
It works by pushing a job to scan the document and when it is finished it triggers a Firebase event that will trigger an event to refresh the model on the component.

## Component event
```
    #[On('file-scanned')]
    public function documentScanned(string $firebasePath): void
    {
        Log::debug('[ScanneableTrait] documentScanned: ');

        if ($this->firebasePath == $firebasePath) {
            $documentId = $this->document->id;
            $this->document = $this->document->fresh();
            if (is_null($this->document)) {
                $this->dispatch('documentRemoved', ['documentId' => $documentId]);
            }
        }
    }
```    
## Event triggered from javascript when a change happened on Firebase

`Livewire.dispatch('file-scanned', {firebasePath: databasePath});`

It works like a charm when the Scanned status is Ok but when the model gets deleted it fails and never passes through the event function documentScanned.

The only error that I see when this happens is:

> Uncaught (in promise) 
> sendRequest	@	livewire.js?id=cc800bf4:4321
> Promise.then		
> (anonymous)	@	livewire.js?id=cc800bf4:4186
> createAndSendNewPool	@	livewire.js?id=cc800bf4:4182
> (anonymous)	@	livewire.js?id=cc800bf4:4158
> (anonymous)	@	livewire.js?id=cc800bf4:4221
> setTimeout		
> bufferPoolingForFiveMs	@	livewire.js?id=cc800bf4:4220
> add	@	livewire.js?id=cc800bf4:4155
> requestCall	@	livewire.js?id=cc800bf4:4237
> (anonymous)	@	livewire.js?id=cc800bf4:4478
> (anonymous)	@	livewire.js?id=cc800bf4:4426
> handler4	@	livewire.js?id=cc800bf4:8709
> dispatchEvent	@	livewire.js?id=cc800bf4:4675
> dispatchGlobal	@	livewire.js?id=cc800bf4:4645
> (anonymous)	@	firebase-file-upload…?v=1728574480:26048
> onValue	@	firebase-file-upload…?v=1728574480:12102
> (anonymous)	@	firebase-file-upload…?v=1728574480:12909
> exceptionGuard	@	firebase-file-upload…js?v=1728574480:709
> eventListRaise	@	firebase-file-upload…?v=1728574480:10871
> eventQueueRaiseQueuedEventsMatchingPredicate	@	firebase-file-upload…?v=1728574480:10846
> eventQueueRaiseEventsForChangedPath	@	firebase-file-upload…?v=1728574480:10835
> repoOnDataUpdate	@	firebase-file-upload…?v=1728574480:11053
> (anonymous)	@	firebase-file-upload…?v=1728574480:10952
> onDataPush_	@	firebase-file-upload…s?v=1728574480:3701
> onDataMessage_	@	firebase-file-upload…s?v=1728574480:3695
> onDataMessage_	@	firebase-file-upload…s?v=1728574480:2524
> onPrimaryMessageReceived_	@	firebase-file-upload…s?v=1728574480:2518
> (anonymous)	@	firebase-file-upload…s?v=1728574480:2420
> appendFrame_	@	firebase-file-upload…s?v=1728574480:2037
> handleIncomingFrame	@	firebase-file-upload…s?v=1728574480:2085
> mySock.onmessage	@	firebase-file-upload…s?v=1728574480:1984

**Reply by wizzymore** [reply]
You need to provide the error trace of the PHP side not the javascript side as the error happens on the server

**Reply by ccssmania** [reply]
> You need to provide the error trace of the PHP side not the javascript side as the error happens on the server

It doesn't give me any PHP trace; when the Livewire.dispatch('file-scanned', {firebasePath: databasePath}); is triggered I get the  javascript error message and the page is reloaded automatically.

**Reply by wizzymore** [reply]
In the network tab of the browser you will have a request that fails

**Reply by ccssmania** [reply]
It just shows a request with 302 response that explain why I'm getting redirected:
![image](https://github.com/user-attachments/assets/af2ce633-52b4-4292-beba-6e46eefa3b57)

**Reply by ccssmania** [reply]
The only thing I see that differs from the other components rendered is this "scripts" that come in the payload of the request:
![image](https://github.com/user-attachments/assets/3a6d36a2-b58a-41b1-82e1-7c742b9a19fd)
and that one is the component where the Model got deleted

**Reply by ccssmania** [reply]
The only way in which it worked when the Document gets deleted is by using Computed properties:
```
    #[Computed()]
    public function document()
    {
        return Document::find($this->documentId);
    }
```
But my concern now is that each update will trigger a query to the database, which can overwhelm it.

**Reply by wizzymore** [reply]
Did you typecast your class property to be the specific model? If yes try to remove that and see if it works. Another solution would be to not use ->fresh() and to just query it like you did in the computed function, get the document with the `Document::find(...)` query

**Reply by ccssmania** [reply]
Yes, I typecasted the class and removed it, and it didn't work. The problem is not the ->fresh() method since it never arrives there.

**Reply by ccssmania** [reply]
## This is my component:
```
class Row extends Component
{

     /**
     * @var Document
     */
    public $document = null;

    /**
     * The real-time database firebase path.
     * @var string
     */
    public string $firebasePath;

    /**
     * Flag to check if the document is being scanned
     * @var bool
     */
    public bool $isScanning = false;

    /**
     * @var int|null
     */
    public ?int $profileId = null;

    public function render()
    {
        $this->isScanning = $this->document?->scanned == Scanned::BEING_SCANNED;

        return view('livewire.document.row');
    }

    public function mount()
    {
        $this->firebasePath = $this->document?->getFirebaseDatabasePath();
    }

    /**
     * Handles the 'file-scanned' event by updating the document's scanned status.
     *
     * @param string $firebasePath The path of the scanned file in the Firebase database.
     * @return void
     */
    #[On('file-scanned')]
    public function documentScanned(string $firebasePath): void
    {
        Log::debug('[ScanneableTrait] documentScanned: '. $firebasePath);

        if ($this->firebasePath == $firebasePath) {
            $this->document = $this->document?->fresh();
        }
    }
}
```

## row.blade.php
```
<div id="{{ $firebasePath }}" class="@if ($document) doc-list-table-row @endif">
    @if ($document)
        <div class="row">
            <div class="columns small-12 medium-4 padding-0">
                <img style="height: 42px;" src="{{ URL::asset('/img/resources/summary.png') }}">
                <a
                    @if( $document->isDownloadable() )
                        onclick="downloadFile(this)"
                    @else
                        style="cursor: not-allowed; color: #ccc;"
                    @endif
                    class="size-sm-12"
                >
                    {{ $document->title }}
                </a>
            </div>
        </div>
    @endif
</div>
@if ($isScanning)
    @script
        <script>
            initFirebase(@json($firebasePath));
        </script>
    @endscript
@endif
```
**If the status of the document is "Scanning", then a script with Firebase will listen for changes:**
```
import { db } from "./firebase";
import { ref, onValue, remove } from "firebase/database";
import Swal from 'sweetalert2'
/**
 * Add event listener for firebase
 * @param {string} firebasePath - path to firebase database
 */
window.initFirebase = function (firebasePath) {

    const conversationRef = ref(db, firebasePath);
    onValue(conversationRef, (snapshot) => {
        if (snapshot.val()) {
            Livewire.dispatch('file-scanned', {firebasePath: firebasePath});
        }
    });

}
```
**If there is a value then it will trigger the event file-scanned, but when the Document was deleted the component with the deleted document never passed through the listener event function, and I just got redirected:**
![image](https://github.com/user-attachments/assets/b6d5fb65-2a98-4123-aefb-3cd384c99d79)

![image](https://github.com/user-attachments/assets/bf8bbaeb-a1e7-4e85-9a9a-7e4580e00611)

**Reply by wizzymore** [reply]
The problem is that the document is being deleted and then you send an event to the livewire component which still is binded to that document's id and tries to pull it from the DB and fails.

What i recommend is that instead of listening for the 'file-scanned' event on each row, i will instead listen for it in a "master" component that has each individual row and remove it from then and just loop over them. That way you never bind a component to a model that is possibly not there.

I recommend you read this: https://livewire.laravel.com/docs/nesting

And most importantly read this:

> **You might not need a Livewire component**
Before you extract a portion of your template into a nested Livewire component, ask yourself: Does this content in this component need to be "live"? If not, we recommend that you create a simple [Blade component](https://laravel.com/docs/blade#components) instead. Only create a Livewire component if the component benefits from Livewire's dynamic nature or if there is a direct performance benefit.

This also applies to child components, do i really need to create a component that is rendered X amount of times or am i fine with just one and some BLADE components instead.

**Reply by wizzymore** [reply]
This issue is more common then we give it credit to.
If you come from react or vue/svelte/angular probably is easier to explain it like: If you have a list of posts and you need to list them, be able to view them and also delete them you will do components like this: `Posts`, `PostRow`, `Post`. The way you will handle the deletion of a `PostRow` from the `Posts` component is not by deleting it from inside itself ( like you are trying to do ). Instead, you will handle that post deletion logic ( api call AND removal from the UI ) from the `Posts` component that holds the array of posts.

The reason why it would work for a component like `Post` ( let's say it's a specific page where you can see the details of the post ) is because after you delete it you REDIRECT AWAY from that page, so you don't hold the deleted Post or it's id nowhere in memory and you never try to pull it from the DB, which, if you would, it will fail.

**Reply by ccssmania** [reply]
Understood!! Thank you.

**Reply by ccssmania** [reply]
It didn't work even using a parent/master component with a collection of documents. When any of the documents get deleted in the job and the event is fired from Javascript, an update happens before I can remove the deleted document from the collection/array, and it is there when it fails and redirects.

**Reply by wizzymore** [reply]
I am sorry mate but i really can't re-create what is happening in your code base. If you have a Database Collection of models held as a public parameter in a Livewire component the missing rows should just be removed, livewire will only throw a 404 answer if parameter was bound to a single model.

This:
```php
public $users

public function mount()
{
    $this->users = User::query()->get();
}
```
Will work even if you delete ALL of the users present in the database.

This:
```php
public $user;


public function mount()
{
    $this->user = User::query()->first();
}
```
This will fail if you remove the model.

I think now the error is coming from your side, i don't know why there is a failure in your code or why there even is a redirect as this is not livewire behaviour.

**Reply by wizzymore** [reply]
> The only way in which it worked when the Document gets deleted is by using Computed properties:
> 
> ```
>     #[Computed()]
>     public function document()
>     {
>         return Document::find($this->documentId);
>     }
> ```
> 
> But my concern now is that each update will trigger a query to the database, which can overwhelm it.

Coming back to this, do it this way if it really worked. Why? Because computed properties are not "computing" on every request, they are evaluated only when they are accessed, this means that, no, it will not trigger a query on all requests.

Also because it is a computed variable you don't even need to do $this->document->fresh() you just `unset($this->document)` and the next time you access `$this->document` in the same request it will have fresh data anyway as it will need to compute itself again.
